packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -2408,6 +2587,21 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin, Resto
     });
   }
 
+  // MATERIAL BANNER API
+
+  // The _messengerMaterialBanner represents the current MaterialBanner being managed by
+  // the ScaffoldMessenger, instead of the Scaffold.
+  ScaffoldFeatureController<MaterialBanner, MaterialBannerClosedReason>? _messengerMaterialBanner;
+
+  // This is used to update the _messengerMaterialBanner by the ScaffoldMessenger.
+  void _updateMaterialBanner() {
+    setState(() {
+      _messengerMaterialBanner = _scaffoldMessenger!._materialBanners.isNotEmpty
+          ? _scaffoldMessenger!._materialBanners.first
+          : null;
+    });
+  }
```

Comment:
we should check here if 

```dart
`_messengerMaterialBanner == _scaffoldMessenger!._materialBanners.first` 
```

And skip in that case to avoid duplicates based on this commit and the current set up in ScaffoldMessenger._updateScaffolds.

Reply:
Same for Snackbars if that is done here.

Reply:
Right, nice catch! Thoughts on the following snippet?
```dart
  void _updateMaterialBanner() {
    final ScaffoldFeatureController<MaterialBanner, MaterialBannerClosedReason>? messengerMaterialBanner =
    _scaffoldMessenger!._materialBanners.isNotEmpty
        ? _scaffoldMessenger!._materialBanners.first
        : null;

    if (_messengerMaterialBanner == messengerMaterialBanner) {
      setState(() {
        _messengerMaterialBanner = messengerMaterialBanner;
      });
    }
  }
```

Reply:
> This is a stellar PR! 
> See comments below, this should also have a second reviewer take a look, I'll see about recruiting someone in the meantime. :)

Thanks! @Piinks

Reply:
Awesome! This looks good. Is there a test validating this? There should be a test that verifies we don't end up duplicating a currently showing snackbar when we push a banner, and vice versa in order to validate this logic and make sure we don't regress later. I think once we have that, this is ready to land! 🎉

Reply:
@Piinks Had that in the back of my mind, just hadn't gotten around to it. I've added them in [the latest commit](https://github.com/flutter/flutter/pull/81706/commits/68d5eb8989714bfcd51c12d391d85531d58a9ce6).